pkg/controller/nexus/update/updates.go

```diff
@@ -0,0 +1,156 @@
+//
+//     This file is part of Nexus Operator.
+//
+//     Nexus Operator is free software: you can redistribute it and/or modify
+//     it under the terms of the GNU General Public License as published by
+//     the Free Software Foundation, either version 3 of the License, or
+//     (at your option) any later version.
+//
+//     Nexus Operator is distributed in the hope that it will be useful,
+//     but WITHOUT ANY WARRANTY; without even the implied warranty of
+//     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//     GNU General Public License for more details.
+//
+//     You should have received a copy of the GNU General Public License
+//     along with Nexus Operator.  If not, see <https://www.gnu.org/licenses/>.
+
+package update
+
+import (
+	ctx "context"
+	"fmt"
+	"github.com/heroku/docker-registry-client/registry"
+	"github.com/m88i/nexus-operator/pkg/apis/apps/v1alpha1"
+	"github.com/m88i/nexus-operator/pkg/logger"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	communityNexusRegistry     = "https://registry.hub.docker.com"
+	communityNexusRepo         = "sonatype/nexus3"
+	tagParseFailureFormat      = "unable to parse tag \"%s\": %v"
+	unableToCheckUpdatesFormat = "Unable to check for updates: %v"
+)
+
+var (
+	latestMicros = make(map[int]string)
+	// necessary to prevent this package from updating the tags while a caller tries to read from it
+	tagsMutex = sync.RWMutex{}
+
+	log = logger.GetLogger("updater")
+)
+
+// GetLatest returns the most recent image tag within a minor (the "y" in "x.y.z").
+// If the minor was not found, the second return value is false. It is expected callers check this.
+func GetLatest(minor int) (tag string, ok bool) {
+	tagsMutex.RLock()
+	defer tagsMutex.RUnlock()
+	tag, ok = latestMicros[minor]
+	return
+}
+
+// CheckUpdatesLoop checks if there is an update available every "delay" duration.
+// It also receives a client used to query the API server so it can start reconciliations for the controller when a newer tag for a given minor is found.
+func CheckUpdatesLoop(reconcileNexus reconcile.Reconciler, client client.Client, delay time.Duration) {
+	for {
+		time.Sleep(delay)
+		tags, err := getTags()
+		if err != nil {
+			log.Errorf(unableToCheckUpdatesFormat, err)
+			continue
+		}
+
+		updated, err := parseTagsAndUpdate(tags)
+		if err != nil {
+			log.Errorf(unableToCheckUpdatesFormat, err)
+			continue
+		}
+
+		if updated {
+			nexusList, err := getNexusList(client)
+			if err != nil && !errors.IsNotFound(err) {
+				log.Errorf("Unable to list deployed Nexus CRs: %v", err)
+			}
+			for _, deployedNexus := range nexusList.Items {
+				result, err := reconcileNexus.Reconcile(reconcile.Request{NamespacedName: types.NamespacedName{Namespace: deployedNexus.Namespace, Name: deployedNexus.Name}})
```

Comment:
We shouldn't call the reconcile loop manually. We can just update the deployment resource and wait for the reconciliation to happen.

Reply:
This part won't be necessary if the reconcile loop is called periodically no matter what. See the other conversation above :-)